promql/engine.go

```diff
@@ -1359,46 +1359,46 @@ func (ev *evaluator) rangeEvalAgg(ctx context.Context, aggExpr *parser.Aggregate
 	}
 	groups := make([]groupedAggregation, groupCount)
 
-	var k int64
-	var ratio float64
 	var seriess map[uint64]Series
-	switch aggExpr.Op {
-	case parser.TOPK, parser.BOTTOMK, parser.LIMITK:
-		if !convertibleToInt64(param) {
-			ev.errorf("Scalar value %v overflows int64", param)
-		}
-		k = int64(param)
-		if k > int64(len(inputMatrix)) {
-			k = int64(len(inputMatrix))
+	// allParamsMatchCondition returns true if the provided condition holds for every float params.
+	allParamsMatchCondition := func(params []FPoint, constParam bool, condition func(float64) bool) bool {
```

Comment:
While looking at this, I had another thought how we might want to encapsulate the param handling a bit better. Instead of this local function and some other bells and whistles, how about having an `fParams` interface like this:

```Go
type fParams interface  {
	Next(ts int64) float64
	Max() float64
	Min() float64
	HasAnyNaN() bool
}
```

There could be one implementation for the variable param series, and one for the constant parameter. The former would pre-populate internal fields to to be used for `Max`, `Min`, `HasAnyNaN`. The latter would store one float value and return it always (or just do `math.IsNaN` for the const value).

Then you could do many operations in an efficient way without using separate `if` statements everywhere.

Reply:
That's a good idea. It would make the code more readable. I'll implement it as soon as possible.